### ISSUE TITLE
More verbose error messages for (geometry) conversion expressions

### ISSUE THREAD

**Issue by kannes** [qgis/QGIS#32550]
**Feature description.**
Playing with Geometry Generator expressions I often get `Eval Error: Cannot convert to geometry`, e.g. when accidentally passing a nested array to `collect_geometries`.

To find out what is wrong, I have to look at the thing passed to the function and because of the non-descript error message I have to remove everything around it.

Examples:

`make_point('a', 3)` gives me an `Eval Error: Cannot convert 'a' to double`, awesome!

```
collect_geometries(
 array(
  array(
   make_point(1, 3),
   make_point(1, 2)
  )
 )
)
```
gives me a non-descript `Eval Error: Cannot convert to geometry`.

It would be awesome if the message could tell me about the type of object the function got (hm, that would not help here), or even better, print the stringified representation of the object it got.

Looking at https://github.com/qgis/QGIS/blob/4b411e0d1af887a0abba4030899a219d6d07317d/src/core/expression/qgsexpressionutils.h#L315 I see some other functions trying to print their faulty arguments but not the ones dealing with geometries, features or nodes(?). I have no idea if I am even looking at the correct code, I just grepped for "cannot convert to geometry". ;)

If this was a trivial fix, it would be awesome. If not, it might not be worth the effort.

**Comment by m-kuhn** [comment]
Try

```
parent->setEvalErrorString( QStringLiteral( "Cannot convert '%1' to geometry" ).arg( value.toString() ); 
```

**Comment by m-kuhn** [comment]
@kannes are you able to build and test a qgis with this change? I'm currently too short in time to do it myself but would be happy to merge a pre-tested pull request.

**Comment by kannes** [comment]
Nope, I did not manage to get it to work. What ever I try, the resulting binary is not using the text I edited. Are those really the correct strings?

This was my patch:

```
--- qgis/src/core/expression/qgsexpressionutils.h	2019-11-09 20:57:07.729590847 +0100
+++ qgis/src/core/expression/qgsexpressionutils.h	2019-11-09 20:57:07.729590847 +0100
@@ -329,7 +329,7 @@
       if ( value.canConvert<QgsGeometry>() )
         return value.value<QgsGeometry>();
 
-      parent->setEvalErrorString( QStringLiteral( "Cannot convert to geometry" ) );
+      parent->setEvalErrorString( QStringLiteral( "Cannot convert '%1' to geometry" ).arg( value.toString() ) );
       return QgsGeometry();
     }
 
@@ -338,7 +338,7 @@
       if ( value.canConvert<QgsFeature>() )
         return value.value<QgsFeature>();
 
-      parent->setEvalErrorString( QStringLiteral( "Cannot convert to feature" ) );
+      parent->setEvalErrorString( QStringLiteral( "Cannot convert '%1' to feature" ).arg( value.toString() ) );
       return 0;
     }
 
@@ -347,7 +347,7 @@
       if ( value.canConvert<QgsExpressionNode *>() )
         return value.value<QgsExpressionNode *>();
 
-      parent->setEvalErrorString( QStringLiteral( "Cannot convert to node" ) );
+      parent->setEvalErrorString( QStringLiteral( "Cannot convert '%1' to node" ).arg( value.toString() ) );
       return nullptr;
     }
 ```

I even entered random gibberish as text but still got `Eval Error: Cannot convert to geometry`.

https://pastebin.com/Wi5pX4Q0 is the PKGBUILD I used to compile and package.
